sklearn/metrics/_pairwise_distances_reduction/_radius_neighborhood.pyx.tp

```diff
@@ -381,15 +384,26 @@ cdef class EuclideanRadiusNeighbors{{name_suffix}}(RadiusNeighbors{{name_suffix}
             )
             ITYPE_t dist_middle_terms_chunks_size = self.Y_n_samples_chunk * self.X_n_samples_chunk
 
-        self.gemm_term_computer = GEMMTermComputer{{name_suffix}}(
-            datasets_pair.X,
-            datasets_pair.Y,
-            self.effective_n_threads,
-            self.chunks_n_threads,
-            dist_middle_terms_chunks_size,
-            n_features=datasets_pair.X.shape[1],
-            chunk_size=self.chunk_size,
-        )
+        if isinstance(self.datasets_pair, DenseDenseDatasetsPair{{name_suffix}}):
+            self.gemm_term_computer = DenseDenseMiddleTermComputer{{name_suffix}}(
+                datasets_pair.X,
+                datasets_pair.Y,
+                self.effective_n_threads,
+                self.chunks_n_threads,
+                dist_middle_terms_chunks_size,
+                n_features=datasets_pair.X.shape[1],
+                chunk_size=self.chunk_size,
+            )
+        else:
+            self.gemm_term_computer = SparseSparseMiddleTermComputer{{name_suffix}}(
+                datasets_pair.X,
+                datasets_pair.Y,
+                self.effective_n_threads,
+                self.chunks_n_threads,
+                dist_middle_terms_chunks_size,
+                n_features=datasets_pair.X.shape[1],
+                chunk_size=self.chunk_size,
+            )
```

Comment:
This works. Ultimately we can have this logic wrapped in a new classmethod for `MiddleTermComputer` when the CSR × Dense and the Dense × CSR are supported to remove duplication and hide the implementation from this file (making `MiddleTermComputer` the only symbol imported from the currently-named `_gemm_term_computer.pxd.tp` file).

Reply:
Got it, in this scenario is it okay to import `DenseDenseDatasetPair` and `SparseSparseDatasetPair` in `_gemm_term_computer.pyx.tp`?  I avoided cross-importing sub-modules, but it might not be important.

Reply:
For now, I would first focus on having `RadiusNeighbors` (or preferably `ArgKmin` since it's simpler) run on a CSR × CSR case.

This when its works, this logic would be wrapped in (let's say) `MiddleTermComputer.get_for` so as to only import `MiddleTermComputer` and here have:

```python
self.middle_term_computer = MiddleTermComputer.get_for(
    X,
    Y,
    self.effective_n_threads,
    self.chunks_n_threads,
    dist_middle_terms_chunks_size,
    n_features=X.shape[1],
    chunk_size=self.chunk_size,
)
```

Reply:
Ok, so importing the DatasetPair classes inside `_gemm_term_computer.pyx.tp` will be fine then?

Reply:
Normally, `DatasetsPair` should an need not be imported in this case: `X` and `Y` can be passed to `MiddleTermComputer.get_for` in this scope without using the `datasets_pair` attribute (which is not needed in this scope as we have discussed).

Reply:
Ok, so you rather duplicate the dispatch logic from DatasetsPair in MiddleTermComputer, instead of using `isinstance(..., SparseSparseDatasetsPair64)`, correct?